QU test is not thread safe

=========================== short test summary info ============================
403
FAILED tests/test_handlers.py::test_questions_query_handler_populates_the_question_history_from_qm_questions
404
ERROR tests/test_handlers.py::test_questions_query_handler_populates_the_question_history_from_qm_questions
405
=== 1 failed, 199 passed, 5 skipped, 1 warning, 1 error in 103.74s (0:01:43) ===
406
make: *** [Makefile:47: test_coverage] Error 1

tests/test_handlers.py:1047: 
298
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
299
zeroconf/_core.py:763: in close
300
    self.engine.close()
301
zeroconf/_core.py:174: in close
302
    asyncio.run_coroutine_threadsafe(self._async_close(), self.loop).result()
303
/opt/hostedtoolcache/PyPy/3.6.12/x64/lib-python/3/concurrent/futures/_base.py:432: in result
304
    return self.__get_result()
305
/opt/hostedtoolcache/PyPy/3.6.12/x64/lib-python/3/concurrent/futures/_base.py:384: in __get_result
306
    raise self._exception
307
/opt/hostedtoolcache/PyPy/3.6.12/x64/lib-python/3/asyncio/tasks.py:180: in _step
308
    result = coro.send(None)
309
zeroconf/_core.py:154: in _async_close
310
    await self._cache_cleanup_task
311
/opt/hostedtoolcache/PyPy/3.6.12/x64/lib-python/3/asyncio/futures.py:329: in __iter__
312
    return self.result()  # May raise too.
313
/opt/hostedtoolcache/PyPy/3.6.12/x64/lib-python/3/asyncio/futures.py:243: in result
314
    raise self._exception
315
/opt/hostedtoolcache/PyPy/3.6.12/x64/lib-python/3/asyncio/tasks.py:180: in _step
316
    result = coro.send(None)
317
zeroconf/_core.py:141: in _async_cache_cleanup
318
    self.zc.question_history.async_expire(now)
319
zeroconf/_history.py:66: in async_expire
320
    for question, now_known_answers in self._history.items()
321
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
322

323
.0 = <dict_itemiterator object at 0x0000000009084f80>
324

325
>       question
326
        for question, now_known_answers in self._history.items()
327
        if now - now_known_answers[0] > _DUPLICATE_QUESTION_INTERVAL
328
    ]
329
E   RuntimeError: dictionary changed size during iteration
330

331
zeroconf/_history.py:65: RuntimeError
332
------------------------------ Captured log call -------------------------------

--- a/tests/test_handlers.py
+++ b/tests/test_handlers.py
@@ -1022,8 +1022,10 @@ async def test_record_update_manager_add_listener_callsback_existing_records():
     await aiozc.async_close()
 
 
-def test_questions_query_handler_populates_the_question_history_from_qm_questions():
-    zc = Zeroconf(interfaces=['127.0.0.1'])
+@pytest.mark.asyncio
+async def test_questions_query_handler_populates_the_question_history_from_qm_questions():
+    aiozc = AsyncZeroconf(interfaces=['127.0.0.1'])
+    zc = aiozc.zeroconf
     now = current_time_millis()
     _clear_cache(zc)
 
@@ -1044,11 +1046,13 @@ def test_questions_query_handler_populates_the_question_history_from_qm_question
     assert multicast_out is None
     assert zc.question_history.suppresses(question, now, set([known_answer]))
 
-    zc.close()
+    await aiozc.async_close()
 
 
-def test_questions_query_handler_does_not_put_qu_questions_in_history():
-    zc = Zeroconf(interfaces=['127.0.0.1'])
+@pytest.mark.asyncio
+async def test_questions_query_handler_does_not_put_qu_questions_in_history():
+    aiozc = AsyncZeroconf(interfaces=['127.0.0.1'])
+    zc = aiozc.zeroconf
     now = current_time_millis()
     _clear_cache(zc)
 
@@ -1069,17 +1073,19 @@ def test_questions_query_handler_does_not_put_qu_questions_in_history():
     assert multicast_out is None
     assert not zc.question_history.suppresses(question, now, set([known_answer]))
 
-    zc.close()
+    await aiozc.async_close()
 
 
-def test_guard_against_low_ptr_ttl():
+@pytest.mark.asyncio
+async def test_guard_against_low_ptr_ttl():
     """Ensure we enforce a minimum for PTR record ttls to avoid excessive refresh queries from ServiceBrowsers.
 
     Some poorly designed IoT devices can set excessively low PTR
     TTLs would will cause ServiceBrowsers to flood the network
     with excessive refresh queries.
     """
-    zc = Zeroconf(interfaces=['127.0.0.1'])
+    aiozc = AsyncZeroconf(interfaces=['127.0.0.1'])
+    zc = aiozc.zeroconf
     # Apple uses a 15s minimum TTL, however we do not have the same
     # level of rate limit and safe guards so we use 1/4 of the recommended value
     answer_with_low_ttl = r.DNSPointer(
@@ -1116,4 +1122,4 @@ def test_guard_against_low_ptr_ttl():
     incoming_answer_normal = zc.cache.async_get_unique(answer_with_normal_ttl)
     assert incoming_answer_normal.ttl == const._DNS_OTHER_TTL
     assert zc.cache.async_get_unique(good_bye_answer) is None
-    zc.close()
+    await aiozc.async_close()

